processor/factory.go

```diff
@@ -0,0 +1,41 @@
+// Copyright 2019, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package processor
+
+import "github.com/spf13/viper"
+
+// TraceDataProcessorFactory is an interface that builds a new TraceDataProcessor based on
+// some viper.Viper configuration.
+type TraceDataProcessorFactory interface {
+	// GetType gets the type of the TraceDataProcessor created by this factory.
+	GetType() string
+	// NewFromViper takes a viper.Viper config and creates a new TraceDataProcessor.
+	NewFromViper(cfg *viper.Viper) (TraceDataProcessor, error)
```

Comment:
Question:
Why pointer? The config should not be changed, also it is not on the critical path, an extra copy will not have performance implications.

Probably we should have:
```go
NewProcessorFromViper(cfg viper.Viper, next TraceDataProcessor) (TraceDataProcessor, error)
```

Reason to not have a list for next is because this can be called by our config module:
```go
factory.NewProcessorFromViper(cfg, processor.NewMultiTraceDataProcessor(tdps))
```

Also we may provide some helper structs:
```go
type RawTraceDataProcessor {
  ProcessTraceData(ctx contex.Context, td data.TraceData) (context.Context, data.TraceData, error)
}

type ChainedTraceDataProcessor {
  next TraceDataProcessor
  current RawTraceDataProcessor
}

var _ TraceDataProcessor = &ChainedTraceDataProcessor(nil)

func NewChainedTraceDataProcessor(current RawTraceDataProcessor, next TraceDataProcessor) {
  return &ChainedTraceDataProcessor{next: next, current: current}
}

func (ctdp *ChainedTraceDataProcessor) ProcessTraceData(ctx contex.Context, td data.TraceData) error {
  ctx, td, err := current.ProcessTraceData(ctx, td)
  if (err != nil) {
    return err;
  }
  return next.ProcessTraceData(ctx, td)
}
```

Similar to ChainedTraceDataProcessor we can have a ChainedWithObservabilityTraceDataProcessor that also records stats/traces.

Reply:
Its a pointer since thats how the viper library returns viper instances and generally how it handles itself (e.g. https://godoc.org/github.com/spf13/viper#New). I don't see the benefit of requiring us to dereference the viper (although I agree the performance hit is negligible in this case). 
Agree on the not needing a list b/c of the helpers we will have.

Reply:
We agree that performance is not the criteria here. While in principle the configuration shouldn't be changed that can't be enforced just by passing by value since the type contains pointers. If we want to enforce that we need to define a proper interface in which we select the allowed methods. Passing by value can be confusing because things like maps used by Viper will be shared while side effects of some functions (e.g.: `SetConfigType`) won't. Passing it as a pointer makes clear that any change to it will be shared. My preference is to keep it as a pointer (at least for now).

Reply:
Clarification regarding performance: I don't expect value vs pointer discussed here to be noticeable but notice that startup time is an important consideration for oc-service in general.